server/container_create_linux.go

```diff
@@ -1237,25 +1244,92 @@ func (s *Server) addOCIBindMounts(ctx context.Context, ctr ctrfactory.Container,
+func (s *Server) mountImage(ctx context.Context, specgen *generate.Generator, imageVolumesPath string, m *types.Mount) (*oci.ContainerVolume, error) {
+	if m == nil || m.Image == nil || m.Image.Image == "" || m.ContainerPath == "" {
+		return nil, fmt.Errorf("invalid mount specified: %+v", m)
+	}
+
+	log.Debugf(ctx, "Image ref to mount: %s", m.Image.Image)
+	status, err := s.storageImageStatus(ctx, types.ImageSpec{Image: m.Image.Image})
 	if err != nil {
-		return "", "", fmt.Errorf("get storage image status: %w", err)
+		return nil, fmt.Errorf("get storage image status: %w", err)
 	}
 
-	id := status.ID.IDStringForOutOfProcessConsumptionOnly()
-	log.Debugf(ctx, "Image ID to mount: %v", id)
+	if status == nil {
+		// Should not happen because the kubelet ensures the image.
+		return nil, fmt.Errorf("image %q does not exist locally", m.Image.Image)
+	}
+
+	imageID := status.ID.IDStringForOutOfProcessConsumptionOnly()
+	log.Debugf(ctx, "Image ID to mount: %v", imageID)
 
 	options := []string{"ro", "noexec", "nosuid", "nodev"}
-	mountPoint, err := s.Store().MountImage(id, options, mountLabel)
+	mountPoint, err := s.Store().MountImage(imageID, options, "")
 	if err != nil {
-		return "", "", fmt.Errorf("mount storage: %w", err)
+		return nil, fmt.Errorf("mount storage: %w", err)
 	}
-
 	log.Infof(ctx, "Image mounted to: %s", mountPoint)
-	return mountPoint, id, nil
+
+	const overlay = "overlay"
+	specgen.AddMount(rspec.Mount{
+		Type:        overlay,
+		Source:      overlay,
+		Destination: m.ContainerPath,
+		Options: []string{
+			"lowerdir=" + mountPoint + ":" + imageVolumesPath,
+		},
+		UIDMappings: getOCIMappings(m.UidMappings),
+		GIDMappings: getOCIMappings(m.GidMappings),
+	})
+	log.Debugf(ctx, "Added overlay mount from %s to %s", mountPoint, imageVolumesPath)
+
+	return &oci.ContainerVolume{
+		ContainerPath:     m.ContainerPath,
+		HostPath:          mountPoint,
+		Readonly:          m.Readonly,
+		RecursiveReadOnly: m.RecursiveReadOnly,
+		Propagation:       m.Propagation,
+		SelinuxRelabel:    m.SelinuxRelabel,
+		Image:             &types.ImageSpec{Image: imageID},
+	}, nil
+}
+
+func (s *Server) ensureImageVolumesPath(ctx context.Context, mounts []*types.Mount) (string, error) {
+	// Check if we need to anything at all
+	noop := true
+	for _, m := range mounts {
+		if m.Image != nil && m.Image.Image != "" {
+			noop = false
+			break
+		}
+	}
+
+	if noop {
+		return "", nil
+	}
+
+	imageVolumesPath := filepath.Join(filepath.Dir(s.Config().ContainerExitsDir), "image-volumes")
+	log.Debugf(ctx, "Using image volumes path: %s", imageVolumesPath)
+
+	if err := os.MkdirAll(imageVolumesPath, 0o700); err != nil {
+		return "", fmt.Errorf("create image volumes path: %w", err)
+	}
+
+	f, err := os.Open(imageVolumesPath)
+	if err != nil {
+		return "", fmt.Errorf("open image volumes path %s: %w", imageVolumesPath, err)
+	}
+
+	_, readErr := f.Readdirnames(1)
```

Comment:
Nit: you can already use the most performant API here.

```suggestion
	_, readErr := f.ReadDir(1)
```

Reply:
Fixed as suggested, thank you!